internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,286 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionstore
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/metrics"
+	"github.com/tikv/client-go/v2/util"
+)
+
+// PipelinedMemDB is a Store which contains
+//   - a mutable buffer for read and write
+//   - an immutable onflushing buffer for read
+//   - like MemDB, PipelinedMemDB also CANNOT be used concurrently
+type PipelinedMemDB struct {
+	// Like MemDB, this RWMutex only used to ensure memdbSnapGetter.Get will not race with
+	// concurrent memdb.Set, memdb.SetWithFlags, memdb.Delete and memdb.UpdateFlags.
+	sync.RWMutex
+	onFlushing        atomic.Bool
+	errCh             chan error
+	flushFunc         FlushFunc
+	bufferBatchGetter BufferBatchGetter
+	memBuffer         *MemDB
+	flushingMemBuffer *MemDB // the flushingMemBuffer is not wrapped by a mutex, because there is no data race in it.
+	len, size         int    // len and size records the total flushed and onflushing memdb.
+}
+
+const (
+	// MinFlushKeys is the minimum number of keys to trigger flush.
+	// small batch can lead to poor performance and resource waste in random write workload.
+	// 10K batch size is large enough to get good performance with random write workloads in tests.
+	MinFlushKeys = 10000
+	MinFlushSize = 16 * 1024 * 1024  // 16MB
+	MaxFlushSize = 128 * 1024 * 1024 // 128MB
+)
+
+type pipelinedMemDBSkipRemoteBuffer struct{}
+
+var pipelinedMemDBSkipRemoteBufferKey = pipelinedMemDBSkipRemoteBuffer{}
+
+// WithPipelinedMemDBSkipRemoteBuffer is used to skip reading remote buffer for saving RPC.
+func WithPipelinedMemDBSkipRemoteBuffer(ctx context.Context) context.Context {
+	return context.WithValue(ctx, pipelinedMemDBSkipRemoteBufferKey, struct{}{})
+}
+
+type FlushFunc func(*MemDB) error
+type BufferBatchGetter func(ctx context.Context, keys [][]byte) (map[string][]byte, error)
+
+func NewPipelinedMemDB(bufferBatchGetter BufferBatchGetter, flushFunc FlushFunc) *PipelinedMemDB {
+	return &PipelinedMemDB{
+		memBuffer:         newMemDB(),
+		errCh:             make(chan error, 1),
+		flushFunc:         flushFunc,
+		bufferBatchGetter: bufferBatchGetter,
+	}
+}
+
+// Dirty returns whether the pipelined buffer is mutated.
+func (p *PipelinedMemDB) Dirty() bool {
+	return p.memBuffer.Dirty() || p.len > 0
+}
+
+// GetMemDB returns the mutable memdb ONLY for writing.
+func (p *PipelinedMemDB) GetMemDB() *MemDB {
+	return p.memBuffer
+}
+
+// Get the value by given key, it returns tikverr.ErrNotExist if not exist.
+// The priority of the value is MemBuffer > flushingMemBuffer > Snapshot.
+func (p *PipelinedMemDB) Get(ctx context.Context, k []byte) ([]byte, error) {
+	v, err := p.memBuffer.Get(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
```

Comment:
If the returned error is not `NotFound`, the error should be returned direcrly, perhaps
```
if err != nil {
   if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
      ...
   } else {
     return nil, err
   }
}
v, err = p.flushingMemBuffer.Get(k)
if err != nil {
   if tikverr.IsErrNotFound(err) && ctx.Value(pipelinedMemDBSkipRemoteBufferKey) == nil {
   }
}
```

is more clear and the error handing would not be forgotten.